data_structures/binary_trees/segment_tree.c

```diff
@@ -0,0 +1,231 @@
+{
+    size_t elem_size = tree->elem_size;
+    memcpy(res, tree->identity, elem_size);
+    elem_size = tree->elem_size;
+    char *root = (char *)tree->root;
+    l += tree->length - 1;
+    r += tree->length - 1;
+    while (l <= r)
+    {
+        if (!(l & 1))
+        {
+            tree->combine(res, root + l * elem_size, res);
+        }
+        if (r & 1)
+        {
+            tree->combine(res, root + r * elem_size, res);
+        }
+        r = (r >> 1) - 1;
+        l = (l >> 1);
+    }
+}
+
+/**
+ * Initializes Segment Tree
+ * Accquires memory for segment tree
+ * and fill the leaves of segment tree with data from array
+ * @param arr the array data upon which segment tree is build
+ * @param elem_size size of each element in segment tree
+ * @param len total no of elements in array
+ * @param identity the identity element for combine_function
+ * @param func the combine_function used to build segment tree
+ *
+ * @returns pointer to sgement tree build
+ */
+segment_tree *segment_tree_init(void *arr, size_t elem_size, size_t len,
+                                void *identity, combine_function func)
+{
+    segment_tree *tree = malloc(sizeof(segment_tree));
+    tree->elem_size = elem_size;
+    tree->length = len;
+    tree->combine = func;
+    tree->root = malloc(sizeof(char) * elem_size * (2 * len - 1));
+    tree->identity = malloc(sizeof(char) * elem_size);
+    char *ptr = (char *)tree->root;
+    memset(ptr, 0, (len - 1) * elem_size);  // Initializing memory
+    ptr = ptr + (len - 1) * elem_size;
+    memcpy(ptr, arr, elem_size * len);  // copy the leaf nodes i.e. array data
+    memcpy(tree->identity, identity, elem_size);  // copy identity element
+    return tree;
+}
+
+/**
+ * Dispose Segment Tree
+ * Frees all heap memory accquired by segment tree
+ * @param tree pointer to segment tree
+ */
+void segment_tree_dispose(segment_tree *tree)
+{
+    free(tree->root);
+    free(tree->identity);
+}
+
+/**
+ * Prints the data in segment tree
+ * The data should be of int type
+ * A utility to print segment tree
+ * with data type of int
+ * @param tree pointer to segment tree
+ */
+void segment_tree_print_int(segment_tree *tree)
+{
+    char *base = (char *)tree->root;
+    size_t i = 0;
+    for (; i < 2 * tree->length - 1; i++)
+    {
+        printf("%d ", *(int *)(base + i * tree->elem_size));
+    }
+    printf("\n");
+}
+
+/**
+ * Utility for test
+ * A function compare for minimum between two integers
+ * This function is used as combine_function for RMQ
+ * @param a pointer to integer a
+ * @param b pointer to integer b
+ * @param c pointer where minimum of a and b is tored as result
+ */
+void minimum(const void *a, const void *b, void *c)
+{
+    *(int *)c = *(int *)a < *(int *)b ? *(int *)a : *(int *)b;
+}
+
+/**
+ * Test RMQ
+ * Testing Segment tree using
+ * Range Minimum Queries
+ */
+static void test()
+{
+    int arr[10] = {1, 0, 3, 5, 7, 2, 11, 6, -2, 8};
```

Comment:
since you are using a specific int-type, better to be explicit about it:
```suggestion
    int32_t arr[10] = {1, 0, 3, 5, 7, 2, 11, 6, -2, 8};
```
this might need `#include <inttypes.h>` in the beginning of the program